core/src/main/java/smile/regression/ElasticNet.java

```diff
@@ -0,0 +1,182 @@
+        this(x, y, lambda1, lambda2, 1E-4, 1000);
+    }
+
+    /**
+     * Constructor. Learn the Elastic Net regularized least squares model.
+     * 
+     * @param x
+     *            a matrix containing the explanatory variables. NO NEED to include
+     *            a constant column of 1s for bias.
+     * @param y
+     *            the response values.
+     * @param lambda1
+     *            the shrinkage/regularization parameter for L1
+     * @param lambda2
+     *            the shrinkage/regularization parameter for L2
+     * @param tol
+     *            the tolerance for stopping iterations (relative target duality
+     *            gap).
+     * @param maxIter
+     *            the maximum number of IPM (Newton) iterations.
+     */
+    public ElasticNet(double[][] x, double[] y, double lambda1, double lambda2, double tol, int maxIter) {
+        if (lambda1 <= 0) {
+            throw new IllegalArgumentException("Please use Ridge instead, wrong L1 portion setting:" + lambda1);
+        }
+        if (lambda2 <= 0) {
+            throw new IllegalArgumentException("Please use LASSO instead, wrong L2 portion setting:" + lambda2);
+        }
+
+        this.lambda1 = lambda1;
+        this.lambda2 = lambda2;
+        this.c = 1 / Math.sqrt(1 + lambda2);
+        this.p = x[0].length;
+        lasso = new LASSO(getAugmentedData(x), getAugmentedResponse(y), this.lambda1 * c, tol, maxIter);
+
+        w = new double[lasso.coefficients().length];
+        double rescale = (1 / c);
+        for (int i = 0; i < w.length; i++) {
+            w[i] = rescale * lasso.coefficients()[i];
+        }
+        b = rescale * lasso.intercept();
+    }
+
+    @Override
+    public double predict(double[] x) {
+        if (x.length != p) {
+            throw new IllegalArgumentException(
+                    String.format("Invalid input vector size: %d, expected: %d", x.length, p));
+        }
+
+        return Math.dot(x, w) + b;
+    }
+
+    public double[] coefficients() {
+        return w;
+    }
+
+    public LASSO lasso() {
+        return lasso;
+    }
+
+    /**
+     * transform the original response array by padding 0 at the tail
+     * 
+     * @param y
+     *            original response array
+     * @return response array with padding 0 at tail
+     */
+    private double[] getAugmentedResponse(double[] y) {
+        double[] ret = new double[y.length + p];
+        System.arraycopy(y, 0, ret, 0, y.length);
+        return ret;
+    }
+
+    /**
+     * transform the original data array by padding a weighted identity matrix and
+     * multiply a scaling
+     * 
+     * @param x
+     *            the original data array
+     * @return data with padding
+     */
+    private double[][] getAugmentedData(double[][] x) {
+        double[][] ret = new double[x.length + p][p];
+        double padding = c * Math.sqrt(lambda2);
+        for (int i = 0; i < ret.length; i++) {
+            if (i <= x.length - 1) {
```

Comment:
It is faster if we split this loop over into two. for (i = 0; i < x.length) and (i=x.length; i < ret.length).